### PR TITLE
refactor: move redirect handling into deno_graph

### PR DESCRIPTION
**Pull request by dsherret**
See explanation in https://github.com/denoland/deno_graph/pull/451

### CODE REVIEW

File: cli/cache/mod.rs [hunk 1]
```diff
@@ -234,29 +236,41 @@ impl Loader for FetchCacher {
         LoaderCacheSetting::Only => Some(CacheSetting::Only),
       };
       file_fetcher
-        .fetch_with_options(FetchOptions {
-          specifier: &specifier,
-          permissions,
-          maybe_accept: None,
-          maybe_cache_setting: maybe_cache_setting.as_ref(),
-          maybe_checksum: options.maybe_checksum,
+        .fetch_once_with_options(FetchOnceOptions {
```
Comment (dsherret) [inline cli/cache/mod.rs:239]: This is the main change. The CLI is now doing a single fetch without following redirects and telling deno_graph about the redirect.

File: tests/specs/info/multiple_redirects/main.out [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+      "emit": null,
+      "map": null,
+      "size": 27,
+      "mediaType": "JavaScript",
+      "specifier": "http://localhost:4545/subdir/redirects/redirect1.js"
+    }
+  ],
+  "redirects": {
+    "http://localhost:4546/subdir/redirects/redirect1.js": "http://localhost:4545/subdir/redirects/redirect1.js",
+    "http://localhost:4548/subdir/redirects/redirect1.js": "http://localhost:4546/subdir/redirects/redirect1.js"
```
Comment (dsherret) [inline tests/specs/info/multiple_redirects/main.out:48]: We store all the redirects in the graph now.

File: cli/Cargo.toml [hunk 3]
```diff
@@ -66,17 +66,17 @@ deno_ast = { workspace = true, features = ["bundler", "cjs", "codegen", "proposa
 deno_cache_dir = { workspace = true }
 deno_config = "=0.15.0"
 deno_core = { workspace = true, features = ["include_js_files_for_snapshotting"] }
-deno_doc = { version = "=0.124.0", features = ["html"] }
-deno_emit = "=0.39.1"
-deno_graph = { version = "=0.72.0", features = ["tokio_executor"] }
+deno_doc = { version = "=0.125.0", features = ["html"] }
+deno_emit = "=0.40.0"
+deno_graph = { version = "=0.73.1", features = ["tokio_executor"] }
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/Cargo.toml:71]: I think we should keep it here to ensure they stay only used in this crate. deno_ast is used in deno_runtime

File: cli/args/mod.rs [hunk 4]
```diff
@@ -185,6 +185,7 @@ pub fn ts_config_to_transpile_and_emit_options(
       jsx_fragment_factory: options.jsx_fragment_factory,
       jsx_import_source: options.jsx_import_source,
       precompile_jsx,
+      precompile_jsx_skip_elements: None,
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/args/mod.rs:188]: Yeah in a separate pr.

File: cli/cache/mod.rs (hunk 1 above)
Comment (dsherret) [reply cli/cache/mod.rs:239]: `fetch_once` was the existing naming. I can rename it to `fetch_no_redirects`

File: cli/file_fetcher.rs [hunk 5]
```diff
@@ -280,51 +307,33 @@ impl FileFetcher {
...
+  ) -> Result<FileOrRedirect, AnyError> {
+    debug!(
+      "FileFetcher::fetch_remote_once() - specifier: {}",
+      specifier
+    );
 
     if self.should_use_cache(specifier, cache_setting) {
-      match self.fetch_cached(specifier, maybe_checksum.clone(), redirect_limit)
+      if let Some(file_or_redirect) =
+        self.fetch_cached_once(specifier, maybe_checksum)?
```
[the remark this answers is not in the document]
Comment (dsherret) [reply cli/file_fetcher.rs:324]: No, this no longer handles redirects. It returns back if a redirect occurred.

**Review by bartlomieju — APPROVED** [review]
LGTM!

**Pull request merged**